src/visualizers/panels/ExecutionIndex/ExecutionIndexControl.js

```diff
@@ -38,45 +41,88 @@ define([
     };
 
     ExecutionIndexControl.prototype._initWidgetEventHandlers = function () {
-        this._widget.setExecutionDisplayed = this.setExecutionDisplayed.bind(this);
+        this._widget.setDisplayedExecutions = this.setDisplayedExecutions.bind(this);
     };
 
-    ExecutionIndexControl.prototype.setExecutionDisplayed = function (id, bool) {
-        var lines = this._linesForExecution[id] || [],
-            otherLines,
-            wasMultiLine = this.displayedExecCount() > 1,
-            isMultiLine;
+    ExecutionIndexControl.prototype.setDisplayedExecutions = function (beforeClickIds, afterClickIds) {
+        let addedExecutions = afterClickIds.filter(id => !beforeClickIds.includes(id));
+        let removedExecutions = beforeClickIds.filter(id => !afterClickIds.includes(id));
+        let added = addedExecutions.length > 0;
+        let updatedExecutions = added ? addedExecutions : removedExecutions;
+        this._updateGraphData(...updatedExecutions, added);
+    };
 
-        this._logger.info(`setting execution ${id} to ${bool ? 'displayed' : 'hidden'}`);
+    ExecutionIndexControl.prototype._updateGraphData = function (id, bool) {
         this.displayedExecutions[id] = bool;
-
-        // If we just crossed the multi line threshold, then update all the lines
-        isMultiLine = this.displayedExecCount() > 1;
-        if (isMultiLine !== wasMultiLine) {
-            // Refresh the other lines visible
-            otherLines = Object.keys(this.displayedExecutions)
-                .filter(eId => this.displayedExecutions[eId] && (eId !== id))
-                .map(id => this._linesForExecution[id] || [])
-                .reduce((l1, l2) => l1.concat(l2), []);
-
-            this._updateLines(otherLines, false);
-            this._updateLines(otherLines, true);
+        let displayedGraphExecIds = Object.keys(this.displayedExecutions)
+            .filter((id) => !!this.displayedExecutions[id]);
+        if (this._currentPlotsDataId) {
+            this._widget.removeNode(this._currentPlotsDataId);
+        }
+        if ( this.displayedExecCount() > 0) {
+            this._consolidateGraphData(displayedGraphExecIds);
         }
-
-        this._updateLines(lines, bool);
     };
 
-    ExecutionIndexControl.prototype._updateLines = function (lines, added) {
-        var action = added ? 'addNode' : 'removeNode';
+    ExecutionIndexControl.prototype._consolidateGraphData = function (graphExecIDs) {
+        let graphIds = graphExecIDs.map(execId => this._graphsForExecution[execId]);
+        let graphDescs = graphIds.map(id => this._getObjectDescriptor(id)).filter(desc => !!desc);
+        if(graphDescs.length > 0){
+            let consolidatedDesc = this._combineGraphDesc(graphDescs);
+            this._currentPlotsDataId = consolidatedDesc.id;
+            let plotlyJSON = this._plotlyDescExtractor.descToPlotlyJSON(consolidatedDesc);
+            plotlyJSON.type = 'graph';
+            this._widget.addNode(plotlyJSON);
```

Comment:
As a minor comment, I expected this to be a method that simply consolidated the graph data and returned the consolidated JSON (as the name suggests). I was surprised to see that it has the side effect of actually setting the node in the widget.

It might be nice to have a clearer separation of the responsibilities of this method and move the call to `addNode` and setting `_currentPlotsDataId` to be outside of the method.